### ISSUE TITLE
Make smart-models a library with a dependency on smart-learner.

### ISSUE THREAD

**Issue by MarcCote** [SMART-Lab/models#13]
We could more easily import reusable pieces of code that are related only to the models (e.g. WeightInitializer, Layer objects, etc.) if smart-models is a library i.e. accessible via PYTHONPATH.

We already started, just make it clean.

**Comment by ASalvail** [comment]
Or we could put those objects in smartpy?

**Comment by MarcCote** [comment]
@mgermain raised a good point about that. We should keep in _smart-_**_learner**_ things only related to the training/learning process. For instance, the `Layer` class should not be in _smart-learner_ since the latter doesn't need to know about it to train a model.

A good rule of thumb to know if something should go in _smart-models_ might be: once a model is trained we should be able to **use** it without having the _smart-learner_ library installed.

That being said, `WeightInitializer` would go in the _smart-learner_ library.

**Comment by MarcCote** [comment]
I think the main focus of _smart-models_ was really fuzzy from the start. We should use this library to put everything that can help us creating models but **no** models _per se_. For instance, we could have many types of layers, activation functions, etc.

That said, we definitively need a repo (yes, yet another one) to keep ours models. We could call it _smartlearner-examples_ or _smartlab-projects_ or something along those lines. That repo will contain one folder per project/example (as we wanted to do initially).

So, we probably don't want a dependency with _smart-learner_.

Also, @ASalvail I think we both agree we don't want to spend time reimplementing every "brick" that goes into _smartmodels_, and prefer reusing stuff from _blocks_ whenever possible. If we do go all the way with _blocks_, we might want to rename this repo _smartmodels-blocks_ just so it's clear to everyone that it has a hard dependency on _blocks_.

**Comment by ASalvail** [comment]
> I think the main focus of smart-models was really fuzzy from the start. We should use this library to put everything that can help us creating models but no models per se. For instance, we could have many types of layers, activation functions, etc.

If we do that, it should be merged in SL.

> That said, we definitively need a repo (yes, yet another one) to keep ours models. We could call it smartlearner-examples or smartlab-projects or something along those lines.

At this point, these would be private repositories.

> So, we probably don't want a dependency with smart-learner.

That would be difficult to achieve, but I guess not impossible. Still, I'd rather merge the models parts back inside SL. You would have no reasons not to.

> If we do go all the way with blocks, we might want to rename this repo smartmodels-blocks just so it's clear to everyone that it has a hard dependency on blocks.

I don't see why. We never though of calling SM _smartmodels-theano_ even though it has a pretty hard dependency on it.

Blocks is still in its infancy. I think we'd rather try to contribute actively to it instead of go on using SM. However, their pipeline for training is really not what I'm looking for, so SL still has something going for it.

**tl;dr** : ditch _smartmodels_ and integrate the good parts into _smartlearners_. Use private repositories for models. Cannonical models could be brought into _smartlearners_, eventually. For now it doesn't really matter as we all work on separate models.

**Comment by MarcCote** [comment]
> tl;dr : ditch smartmodels and integrate the good parts into smartlearners. 

Right now, there is nothing in _smartmodels_ excepts the perceptron model. What would you want to integrate in _smartlearners_?

> Use private repositories for models. Cannonical models could be brought into smartlearners, eventually. For now it doesn't really matter as we all work on separate models.

I'll create a _smartlearner-examples_ repo. I intend to fork it and have my different models I'm working on in different branches. Feel free to use your own private repo (my student pack expired so I can't have private repo on github anymore).

**Comment by ASalvail** [comment]
> Right now, there is nothing in smartmodels excepts the perceptron model. What would you want to integrate in smartlearners?

Most of the content of PR #15 

> (my student pack expired so I can't have private repo on github anymore)

That's weird : Github just gave me 5 private repos for life.
